Make MOOSE compliant `--with-strict-petscerrorcode`

lindsayad opened the pull request:
Closes #27784

lindsayad committed:
Make MOOSE compliant `--with-strict-petscerrorcode`

Closes #27784

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2260285/) on 85398d8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 843ddc9ad7abe9002a0444fb8bf3dce129299805`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2260361/) on 94e3825 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27785/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/2260357/) on 94e3825 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2260376/) on 94e3825 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.09%</td>
      <td>85.07%</td>
      <td>-0.02%</td>
      <td>61.27%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104028</td>
      <td>104076</td>
      <td>+48</td>
      <td>174</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18226</td>
      <td>18266</td>
      <td>+40</td>
      <td>110</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.45%</td>
      <td>90.45%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4932</td>
      <td>4936</td>
      <td>+4</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>521</td>
      <td>521</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/contact)

### External petsc solver

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>86.25%</td>
      <td>+0.85%</td>
      <td>95.29%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>351</td>
      <td>389</td>
      <td>+38</td>
      <td>81</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>60</td>
      <td>62</td>
      <td>+2</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/external_petsc_solver)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/external_petsc_solver)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.18%</td>
      <td>84.18%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17034</td>
      <td>17037</td>
      <td>+3</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3201</td>
      <td>3201</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/navier_stokes)

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.61%</td>
      <td>88.63%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1976</td>
      <td>1980</td>
      <td>+4</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>254</td>
      <td>254</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/optimization)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>843ddc</th>
      <th colspan=4>#27785 94e382</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.62%</td>
      <td>90.61%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7928</td>
      <td>7922</td>
      <td>-6</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>821</td>
      <td>821</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27785/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27785/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 61.27% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -966,16 +966,16 @@ DMCreateFieldDecomposition_Moose(
-          ierr = MPI_Scan(&len,
-                          &off,
-                          1,
+          ierr = cast_int<PetscErrorCode>(MPI_Scan(&len,
```
Wait, what are we doing here?  MPI_Scan doesn't return a PETSc error code.

roystgnr replied:
To be sure, the MPI standard mandates `MPI_SUCCESS==0`, so unless the scan actually fails we're fine, but if we see a failure the error message is going to be seriously confusing.

roystgnr commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -140,7 +140,7 @@ DMMooseValidityCheck(DM dm)
                      ((PetscObject)dm)->type_name,
                      DMMOOSE);
-  PetscFunctionReturn(0);
+  return PetscErrorCode(0);
```
Should we just be doing `return PETSC_SUCCESS` for all these?  More self-documenting IMHO.

lindsayad replied:
That works for me because I do not care to be compliant with old PETScs. I was reticent to do this in libMesh because I figure you and @jwpeterson may want to be more strict about being compatible with old versions. `PETSC_SUCCESS` was first introduced in 2022 I believe

roystgnr replied:
2022?  Whoa; I didn't realize that.

roystgnr replied:
I'm fine either way here, then.

roystgnr commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -992,38 +1004,40 @@ attachCallbacksToMat(EigenProblem & eigen_problem, Mat mat, bool eigen)
-                             "formJacobian",
-                             eigen ? Moose::SlepcSupport::mooseSlepcEigenFormJacobianB
-                                   : Moose::SlepcSupport::mooseSlepcEigenFormJacobianA);
+  LIBMESH_CHKERR(
```
This is an interesting idiom for it.  Any reason for preferring this here vs a named `ierr` elsewhere?

roystgnr reviewed (approved):
This is all an improvement over "just hope `ierr==PETSC_SUCCESS`", for sure.

lindsayad committed:
Address Roy review

- Use `PETSC_SUCCESS`
- Don't convert result of `MPI_Scan` to `PetscErrorCode`
- Be consistent about error checking idioms in files

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2262907/) on a1b07a1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27785/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27785/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d1054172417ebf334d27e0155532dada65dcff71`

lindsayad commented:
failures are unrelated
